File: java/JaeHyeon/pro_87946.java
class Solution {
    static int answer = 0;
    static boolean[] visit;

    public int solution(int k, int[][] dungeons) {

        visit = new boolean[dungeons.length];

        dfs(k, dungeons, 0);

        return answer;
    }

    static private void dfs(int fatigue, int[][] dungeon, int depth) {

        for (int i = 0; i < dungeon.length; i++) {

            if (!visit[i] && fatigue >= dungeon[i][0]) {
                visit[i] = true;
                dfs(fatigue-dungeon[i][1], dungeon, depth+1);
                visit[i] = false;
            }
        }

        answer = Math.max(answer, depth);
    }
}

File: soya/src/pro_87946.ts
function solution(k, dungeons) {
  // a:최소 피로도, b:소모 피로도
  let answer = -1;
  let visited = new Array(dungeons.length).fill(false);

  function checkAll(remain, count) {
    //remain: 남은 피로도
    if (count >= dungeons.length) {
      answer = count > answer ? count : answer;
      return;
    }
    for (let i = 0; i < dungeons.length; i++) {
      let dungeon = dungeons[i];
      //방문한적없음 && 피로도 방문 가능
      if (!visited[i] && remain >= dungeon[0]) {
        visited[i] = true;
        checkAll(remain - dungeon[1], count + 1);
        visited[i] = false;
      }
    }
    answer = count > answer ? count : answer;
  }
  checkAll(k, 0);

  return answer;
}
